clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -51,7 +54,13 @@
     private final Selectable selector;
     
     private final MetadataUpdater metadataUpdater;
-
+    
+    /* a list of nodes we've connected to in the past */
+    private List<Integer> nodesEverSeen;
+    private Map<Integer, Node> nodesEverSeenById;
+    /* random offset into nodesEverSeen list */
+    private Random randOffset;
```

Comment:
Thanks for changing this. A couple more minor suggestions:
* We can use this `Random` in the assignment of this.nodeIndexOffset (creating a `Random` without a seed involves a `System.nanoTime` and an atomic update of a JVM global variable).
* All the fields that are assigned in the constructor should be made `final`.